File: udyam/lib/screens/scan_invoice_screen/models/invoice.dart
double? _asDouble(dynamic value) {
  if (value is num) return value.toDouble();
  return double.tryParse(value?.toString() ?? '');
}

class InvoiceSupplier {
  const InvoiceSupplier({this.name, this.phone});

  final String? name;
  final String? phone;

  factory InvoiceSupplier.fromJson(Map<String, dynamic>? json) {
    return InvoiceSupplier(
      name: json?['name']?.toString(),
      phone: json?['phone']?.toString(),
    );
  }
}

class InvoiceItem {
  const InvoiceItem({
    required this.name,
    this.quantity,
    this.unit,
    this.unitPrice,
    this.total,
  });

  final String name;
  final double? quantity;
  final String? unit;
  final double? unitPrice;
  final double? total;

  factory InvoiceItem.fromJson(Map<String, dynamic> json) {
    return InvoiceItem(
      name: json['name']?.toString().trim() ?? '',
      quantity: _asDouble(json['quantity']),
      unit: json['unit']?.toString(),
      unitPrice: _asDouble(json['unit_price']),
      total: _asDouble(json['total']),
    );
  }

  Map<String, dynamic> toInventoryJson({
    String? supplierName,
    String? invoiceNumber,
  }) {
    return {
      'name': name,
      'quantity': quantity ?? 1,
      'unit': unit ?? 'unit',
      'purchase_unit_price': unitPrice,
      'line_total': total,
      'supplier_name': supplierName,
      'invoice_number': invoiceNumber,
    };
  }
}

class Invoice {
  const Invoice({
    this.invoiceNumber,
    this.date,
    required this.supplier,
    required this.items,
    this.subtotal,
    this.tax,
    this.grandTotal,
  });

  final String? invoiceNumber;
  final String? date;
  final InvoiceSupplier supplier;
  final List<InvoiceItem> items;
  final double? subtotal;
  final double? tax;
  final double? grandTotal;

  factory Invoice.fromJson(Map<String, dynamic> json) {
    final rawItems = json['items'];
    return Invoice(
      invoiceNumber: json['invoice_number']?.toString(),
      date: json['date']?.toString(),
      supplier: InvoiceSupplier.fromJson(
        json['supplier'] is Map
            ? Map<String, dynamic>.from(json['supplier'] as Map)
            : null,
      ),
      items: rawItems is List
          ? rawItems
                .whereType<Map>()
                .map(
                  (item) =>
                      InvoiceItem.fromJson(Map<String, dynamic>.from(item)),
                )
                .where((item) => item.name.isNotEmpty)
                .toList()
          : const [],
      subtotal: _asDouble(json['subtotal']),
      tax: _asDouble(json['tax']),
      grandTotal: _asDouble(json['grand_total']),
    );
  }
}

File: backend/app/schemas/invoice.py
from datetime import date as Date
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator


class InvoiceSupplier(BaseModel):
    name: str | None = None
    phone: str | None = None

    @field_validator("name", "phone")
    @classmethod
    def strip_optional_text(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None


class InvoiceItem(BaseModel):
    name: str = Field(min_length=1, max_length=300)
    quantity: Decimal | None = None
    unit: str | None = Field(default=None, max_length=60)
    unit_price: Decimal | None = None
    total: Decimal | None = None

    @field_validator("name")
    @classmethod
    def strip_name(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("item name cannot be blank")
        return value

    @field_validator("unit")
    @classmethod
    def strip_unit(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None

    @field_validator("quantity", "unit_price", "total")
    @classmethod
    def finite_decimal(cls, value: Decimal | None) -> Decimal | None:
        if value is not None and not value.is_finite():
            raise ValueError("must be a finite decimal")
        return value


class InvoiceAnalysis(BaseModel):
    invoice_number: str | None = Field(default=None, max_length=120)
    date: Date | None = None
    supplier: InvoiceSupplier = Field(default_factory=InvoiceSupplier)
    items: list[InvoiceItem] = Field(min_length=1)
    subtotal: Decimal | None = None
    tax: Decimal | None = None
    grand_total: Decimal | None = None

    @field_validator("invoice_number")
    @classmethod
    def strip_invoice_number(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        return value or None

    @field_validator("subtotal", "tax", "grand_total")
    @classmethod
    def finite_decimal(cls, value: Decimal | None) -> Decimal | None:
        if value is not None and not value.is_finite():
            raise ValueError("must be a finite decimal")
        return value
